doc/design/block.md

```diff
@@ -0,0 +1,210 @@
+
+  // find a VarDesc by name, if recursive true, find parent's SymbolTable
+  // recursively.
+  // this interface is introduced to support InferShape, find protobuf messages
+  // of variables and operators, pass pointers into InferShape.
+  // operator
+  //
+  // NOTE maybe some C++ classes such as VarDescBuilder and OpDescBuilder should
+  // be proposed and embedded into pybind to enable python operate on C++ pointers.
+  VarDesc* FindVar(const string& name, bool recursive=true);
+
+  OpDesc* FindOp(const string& name);
+
+  BlockDesc Serialize() const;
+  void DeSerialize();
+
+ private:
+  SymbolTable* parent_;
+
+  map<string, OpDesc> ops_;
+  map<string, VarDesc> vars_;
+};
+```
+
+After all the description of variables and operators is added into SymbolTable,
+the block has enough information to run.
+
+To make block simpler, we use another class `RuntimeTable` to create Variable and Operator objects. `RuntimeTable` is free to add memory optimizements in the future.
+
+The `Block` class takes a `BlockDesc` as input, and provide `Run` and `InferShape` functions.
+
+
+```c++
+namespace {
+// A RuntimeTable maintains all the runtime environment symbols of a block,
+// including the variables and operators.
+// NOTE the implementation of runtime environment is seperated from Block
+// to make it easier to improve memory usage and some other improvements
+// based on a block's description.
+// NOTE this name is aweful, and should be updated latter.
+class RuntimeTable {
+ public:
+  // scope: the scope to create new variables.
+  RuntimeTable(const BlockDesc& desc, const Scope& scope);
+
+  vector<OperatorBase*> ops() { return ops_; }
+
+  bool IsValid() const { return ops_.empty(); }
+
+ protected:
+  void CreateVars(const Scope& scope);
+  void CreateOps();
+
+ private:
+  SymbolTable symbol_tabel_;
+  vector<OperatorBase*> ops_;
+};
+}
+
+
+class Block : OperatorBase {
+public:
+  // desc may be serialized from a RuntimeTable or not.
+  Block(const BlockDesc& desc) desc_(desc) {}
+
+  void InferShape(const framework::Scope& scope) const override {
+    // should run InferShape first.
+    if (!runtime_table_.IsValid()) {
+      runtime_table_ = RuntimeTable(desc_, scope);
+    }
+    for (auto& op : runtime_table_.ops()) {
+      op->InferShape(scope);
+    }
+  }
+
+  void Run(const framework::Scope& scope,
+           const platform::DeviceContext& dev_ctx) const override {
+    for (auto& op : runtime_table_.ops()) {
+      op->Run(scope, dev_ctx);
+    }
+  }
+
+  // some other necessary interfaces of NetOp are list below
+  // ...
+
+private:
+  RuntimeTable runtime_table_;
+  BlockDesc desc_;
+};
+```
+
+## Run and Eval targets
+Block inherits from OperatorBase, which has a Run method. 
+Block's Run method will run its operators sequentially.
+
+There is another important interface called `Eval`, which take some arguments called targets, and generate a minimal graph which takes targets as the end points and creates a new Block, 
```

Comment:
Feels like we need a function maybe called `Prune` which generate a subgraph (or maybe block) from a graph. And then call `Run` to run it.

Reply:
good idea, I will add a function called `Prune`